File: Linked List/141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) 
{
    struct ListNode* ptr1 = head;
    struct ListNode* ptr2 = head;
    if (!head)
    {
        return false;
    }
    if (ptr1->next == NULL)
    {
    
        return false;
    }
    ptr2 = ptr2->next->next;
    if (ptr2 == NULL)
    {
        return false;
    }
    
    while (ptr2 != NULL && ptr2->next != NULL)
    {
        if (ptr1!=ptr2)
        {
            ptr1= ptr1->next;
            ptr2= ptr2->next->next;
            
        }
        else 
        {
            return true;
        }
        
        
    }
    return false;
    
}

File: Linked List/141 Linked List Cycle.py
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        #so in order to detect a cycle, we can use floyd's haire and tortoise algorithm
        
        rabbit = head
        tortoise = head
        
        while rabbit != None and rabbit.next != None:
            rabbit = rabbit.next.next
            tortoise = tortoise.next
            if tortoise == rabbit:
                return True
        
        return False